hatgame-backend/server/game/game.go

```diff
@@ -0,0 +1,348 @@
+	}
+	if len(g.Players.WordsByUser[id]) == g.NumWords {
+		return false, fmt.Errorf("words limit reached")
+	}
+	if _, ok := g.Players.Words[word]; ok {
+		return false, nil
+	}
+	g.Players.WordsByUser[id][word] = struct{}{}
+	g.Players.Words[word] = struct{}{}
+	return true, nil
+}
+
+func (g *Game) AddWord(id uint, word string) {
+	ok, err := g.addWord(id, word)
+	if err != nil {
+		g.Errors <- err
+	}
+
+	if !ok {
+		g.Events <- Event{
+			GameID:    g.ID,
+			Type:      message.Error,
+			Msg:       "Already used this word",
+			Receivers: []uint{id},
+		}
+		return
+	}
+
+	g.Events <- Event{
+		GameID:    g.ID,
+		Type:      message.AddWord,
+		Msg:       word,
+		Receivers: []uint{id},
+	}
+
+	if g.CheckWordsFinished() {
+		g.MakeTeams()
+		NotifyGameStarted(g)
+		NotifyStoryteller(g)
+	}
+}
+
+func (g *Game) CheckWordsFinished() bool {
+	g.Players.WordsMutex.RLock()
+	defer g.Players.WordsMutex.RUnlock()
+	return len(g.Players.Words) == g.NumPlayers*g.NumWords
+}
+
+func (g *Game) MakeTeams() {
+	g.Players.WordsMutex.RLock()
+	for id := range g.Players.WordsByUser {
+		g.Process.Teams = append(g.Process.Teams, id)
+	}
+	g.Players.WordsMutex.RUnlock()
+
+	rand.Shuffle(
+		len(g.Process.Teams),
+		func(i, j int) {
+			g.Process.Teams[i], g.Process.Teams[j] = g.Process.Teams[j], g.Process.Teams[i]
+		},
+	)
+}
+
+func NotifyGameStarted(g *Game) {
+	for i, id := range g.Process.Teams {
+		g.Events <- Event{
+			GameID:    g.ID,
+			Type:      message.Team,
+			Msg:       g.Process.Teams[(i+int(float64(g.NumPlayers)/2))%g.NumPlayers],
+			Receivers: []uint{id},
+		}
+	}
+}
+
+func NotifyGameInfo(game *Game) {
+	game.Events <- Event{
+		GameID:    game.ID,
+		Type:      message.GameInfo,
+		Msg:       game,
+		Receivers: game.PlayersIDs,
+	}
+}
+
+func NotifyGameEnded(game *Game) {
+	game.GetResults()
+	game.Events <- Event{
+		GameID:    game.ID,
+		Receivers: game.PlayersIDs,
+		Type:      message.End,
+		Msg:       game.Process.Result,
+	}
+}
+
+func NotifyStoryteller(game *Game) {
+	game.Events <- Event{
+		GameID:    game.ID,
+		Type:      message.Start,
+		Msg:       game.Process.Teams[game.Process.Storyteller],
+		Receivers: game.PlayersIDs,
+	}
+}
+
+func NotifyWord(game *Game, story string) {
+	game.Events <- Event{
+		GameID:    game.ID,
+		Receivers: []uint{game.Process.Teams[game.Process.Storyteller]},
+		Type:      message.Story,
+		Msg:       story,
+	}
+}
+
+func (g *Game) MakeTurn(id uint) {
+	story, found := g.nextWord()
+
+	if !found {
+		NotifyGameEnded(g)
+		return
+	}
+
+	NotifyWord(g, story)
+
+	timer := time.NewTicker(1 * time.Second)
+	timerDone := make(chan struct{})
+	go tick(g, timerDone, timer)
+	go func(timerDone chan struct{}) {
+		time.Sleep(time.Duration(g.Timer) * time.Second)
+		timer.Stop()
+		close(timerDone)
+	}(timerDone)
+}
+
+func tick(game *Game, timerDone chan struct{}, timer *time.Ticker) {
+	i := game.Timer
+
+	for {
+		select {
+		case <-timerDone:
```

Comment:
You can just use time.After and drop timerDone channel.
```suggestion
		case <-time.After(time.Duration(g.Timer) * time.Second):
```

Reply:
This way there is a slight pause between the closing of the ticking channel and the <- time.After because of the creation time. I'm not sure if it's ok.

Reply:
Done